lib/omniauth/strategies/realme.rb

```diff
@@ -18,29 +39,45 @@ def request_phase
         redirect req.create(saml_settings, 'SigAlg' => 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256')
       end
 
-      def callback_phase # rubocop:disable Metrics/AbcSize
+      def callback_phase # rubocop:disable Metrics/AbcSize, Metrics/MethodLength, Metrics/PerceivedComplexity
         response = ::OneLogin::RubySaml::Response.new(request.params['SAMLResponse'],
                                                       settings: saml_settings,
                                                       allowed_clock_drift: allowed_clock_drift)
 
+        ##
+        # If the Realme Context Mapping Service (RCMS) is enabled in Realme
+        # for our app then we will get a RCMS Login Access Token in the
+        # SAMLResponse.
+        #
+        # We save the token if it exists. See
+        # https://developers.realme.govt.nz/how-realme-works/whats-realme-rcms/
+        #
         if response.is_valid?
-          @uid = response.nameid
-          session[:uid] = response.nameid
-
-          ##
-          # If the Realme Context Mapping Service (RCMS) is enabled in Realme
-          # for our app then we will get a RCMS Login Access Token in the
-          # SAMLResponse.
-          #
-          # We save the token if it exists. See
-          # https://developers.realme.govt.nz/how-realme-works/whats-realme-rcms/
-          #
           @realme_cms_lat = response.attributes[RCMS_LAT_NAME] if response.attributes[RCMS_LAT_NAME]
+        end
+
+        if legacy_rails_session_behaviour_enabled?
+          OmniAuth.logger.info "Deprecation: omniauth-realme will stop putting values via Rails `session` in a future version. Use request.env['omniauth.auth'] instead." # rubocop:disable Layout/LineLength
+
+          if response.is_valid?
+            session[:uid] = response.nameid
+          else
+            session[:realme_error] = {
+              error: response.errors.join[/=> (\S+) ->/, 1],
+              message: default_error_messages_for_rails_session(response.errors.join)
+            }
+          end
         else
-          session[:realme_error] = {
-            error: response.errors.join[/=> (\S+) ->/, 1],
-            message: default_error_messages(response.errors.join)
-          }
+          if response.is_valid? # rubocop:disable Style/IfInsideElse
```

Comment:
I don't love the nested if/else stuff I have here. I chose it because it allows all the logic relating to the Rails session stuff to be removed in some future version.